File: Python/greedy-concurrent.py
import copy
import math
import string
import sys
import timeit
import os
from multiprocessing import Pool, Process


class Activity(object):
    def __init__(self, id, lat, lon):
        self.id = id
        self.lat = lat
        self.lon = lon


class Resource(object):
    def __init__(self, id, lat, lon):
        self.id = id
        self.lat = lat
        self.lon = lon


class Allocation(object):
    def __init__(self, rid, aid, dist):
        self.rid = rid
        self.aid = aid
        self.dist = dist


class SchemaData(object):
    activity = []
    resource = []


earthRadiusM = 6367450.0
convert2Rad = math.pi / 180.0
convert2Deg = 180.0 / math.pi
seconds_per_metre = 0.0559234073


def scheduleResources(sd, i):
    allocation = []
    for res in sd.resource:
        for c in range(0, 50):
            lowest = sys.float_info.max
            lowestact = None
            for act in sd.activity:
                dist = distanceBetweenPointsLatLong(
                    res.lat, res.lon, act.lat, act.lon)
                if dist < lowest:
                    lowest = dist
                    lowestact = act
            allocation.append(Allocation(res.id, lowestact.id, lowest))
            sd.activity.remove(lowestact)
    tot = sum(i.dist for i in allocation)
    print(str(i + 1) + ": " + str(tot))


def distanceBetweenPointsLatLong(lat1, lon1, lat2, lon2):
    dStartLatInRad = lat1 * convert2Rad
    dStartLongInRad = lon1 * convert2Rad
    dEndLatInRad = lat2 * convert2Rad
    dEndLongInRad = lon2 * convert2Rad
    dLongitude = dEndLongInRad - dStartLongInRad
    dLatitude = dEndLatInRad - dStartLatInRad
    dSinHalfLatitude = math.sin(dLatitude * 0.5)
    dSinHalfLongitude = math.sin(dLongitude * 0.5)
    a = dSinHalfLatitude * dSinHalfLatitude + \
        math.cos(dStartLatInRad) * math.cos(dEndLatInRad) * \
        dSinHalfLongitude * dSinHalfLongitude
    c = math.atan2(math.sqrt(a), math.sqrt(1.0 - a))
    return earthRadiusM * (c + c)


def run():
    a = []
    r = []
    f = open('../Data/DataSPIF.csv', 'r')
    for line in f:
        items = line.split(',')
        if len(items) == 3:
            r.append(Resource(items[0], float(items[1]), float(items[2])))
        else:
            a.append(Resource(items[0], float(items[1]), float(items[2])))

    with Pool(processes=8) as pool:
        for i in range(0, 1000):
            sdi = SchemaData()
            sdi.resource = list(r)
            sdi.activity = list(a)
            pool.apply_async(scheduleResources, [sdi, i])
        pool.close()
        pool.join()


if __name__ == '__main__':
    run()

File: NodeJS/greedy-concurrent.js
var fs = require('fs');

const earthRadiusM = 6367450;
const convert2Rad = Math.PI / 180.0;
const convert2Deg = 180.0 / Math.PI;
const seconds_per_metre = 0.0559234073;
function distanceBetweenPointsLatLong(lat1, lon1, lat2, lon2) {
  const dStartLatInRad = lat1 * convert2Rad;
  const dStartLongInRad = lon1 * convert2Rad;
  const dEndLatInRad = lat2 * convert2Rad;
  const dEndLongInRad = lon2 * convert2Rad;
  const dLongitude = dEndLongInRad - dStartLongInRad;
  const dLatitude = dEndLatInRad - dStartLatInRad;
  const dSinHalfLatitude = Math.sin(dLatitude * 0.5);
  const dSinHalfLongitude = Math.sin(dLongitude * 0.5);
  const a = dSinHalfLatitude * dSinHalfLatitude + Math.cos(dStartLatInRad) * Math.cos(dEndLatInRad) * dSinHalfLongitude * dSinHalfLongitude;
  const c = Math.atan2(Math.sqrt(a), Math.sqrt(1.0 - a));
  return earthRadiusM * (c + c);
}

function scheduleResources(sd) {
  for (r in sd.resource) {
    const res = sd.resource[r];
    for (let c = 0; c < 50; c++) {
      let lowest = Number.MAX_VALUE;
      let lowestid = '';
      sd.activity.forEach(act => {
        const dist = distanceBetweenPointsLatLong(res.lat, res.lon, act.lat, act.lon);
        if (dist < lowest) {
          lowest = dist;
          lowestid = act.id;
        }
      });
      let a = { resourceId: res.id, activityId: lowestid, distance: lowest };
      sd.allocation.push(a);
      sd.activity.delete(lowestid);
    }
  }
}

// Read in CSV
var contents = fs
  .readFileSync('../Data/DataSPIF.csv', 'utf8')
  .toString()
  .split('\n');
const sd = {
  activity: new Map(),
  resource: [],
  allocation: [],
};
for (i in contents) {
  const line = contents[i];
  const split = line.split(',');
  const lat = parseFloat(split[1]);
  const lon = parseFloat(split[2]);
  switch (split.length) {
    case 3:
      sd.resource.push({ id: split[0], lat, lon });
      break;
    case 4:
      sd.activity.set(split[0], { id: split[0], lat, lon });
      break;
  }
}

// Run schedule routine 100 times
async function go() {
  const m = Array.apply(null, { length: 1000 }).map(Number.call, Number);
  await Promise.all(
    m.map(async i => {
      const sdi = { activity: new Map(sd.activity), resource: sd.resource, allocation: [] };
      scheduleResources(sdi);
      let sum = 0;
      for (a in sdi.allocation) {
        sum += sdi.allocation[a].distance;
      }
      console.log(i + ' : ' + sum);
    }),
  );
}

go();
